go-server-server/main.go

```diff
@@ -53,7 +57,25 @@ func StartHttpsServer(handler http.Handler) {
     }
 
     log.Printf("info: https endpoint started")
-    log.Fatal(server.ListenAndServeTLS(*sw.ServerCertFlag, *sw.ServerKeyFlag))
+
+    // Listening should happen in a go-routine to prevent blocking
+    go func() {
+        log.Fatal(server.ListenAndServeTLS(*sw.ServerCertFlag, *sw.ServerKeyFlag))
+    }()
+
+    sigchannel := make(chan os.Signal, 1)
+    signal.Notify(sigchannel,
+        syscall.SIGTERM,
+        syscall.SIGQUIT)
+
+    <-sigchannel
+    log.Printf("info: Signal received. Shutting down...")
+    if err := server.Shutdown(context.Background()); err != nil {
+        log.Printf("trace: HTTP server Shutdown: %v", err)
+    } else {
+        log.Printf("info: Server shutdown successful!")
+    }
+    os.Exit(0)
```

Comment:
where do you wait till the current job is done?

Reply:
Server.Shutdown() actually waits for all requests to be serviced before shutting it down. https://golang.org/pkg/net/http/#Server.Shutdown